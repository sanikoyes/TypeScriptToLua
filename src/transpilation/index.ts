import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { parseConfigFileWithSystem } from "../cli/tsconfig";
import { CompilerOptions } from "../CompilerOptions";
import { createEmitOutputCollector, TranspiledFile } from "./output-collector";
import { EmitResult, Transpiler } from "./transpiler";

export { Plugin } from "./plugins";
export * from "./transpile";
export * from "./transpiler";
export { EmitHost } from "./utils";
export { TranspiledFile };

export function transpileFiles(
    rootNames: string[],
    options: CompilerOptions = {},
    writeFile?: ts.WriteFileCallback
): EmitResult {
    const program = ts.createProgram(rootNames, options);
    const { diagnostics: transpileDiagnostics, emitSkipped } = new Transpiler().emit({ program, writeFile });
    const diagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...transpileDiagnostics,
    ]);

    return { diagnostics: [...diagnostics], emitSkipped };
}

export function transpileProject(
    configFileName: string,
    optionsToExtend?: CompilerOptions,
    writeFile?: ts.WriteFileCallback
): EmitResult {
    const parseResult = parseConfigFileWithSystem(configFileName, optionsToExtend);
    if (parseResult.errors.length > 0) {
        return { diagnostics: parseResult.errors, emitSkipped: true };
    }

    return transpileFiles(parseResult.fileNames, parseResult.options, writeFile);
}

const libCache: { [key: string]: ts.SourceFile } = {};

/** @internal */
export function createVirtualProgram(input: Record<string, string>, options: CompilerOptions = {}): ts.Program {
    const compilerHost: ts.CompilerHost = {
        fileExists: fileName => fileName in input || ts.sys.fileExists(fileName),
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile() {},

        getSourceFile(fileName) {
            if (fileName in input) {
                return ts.createSourceFile(fileName, input[fileName], ts.ScriptTarget.Latest, false);
            }

            let filePath: string | undefined;

            if (fileName.startsWith("lib.")) {
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                filePath = path.join(typeScriptDir, fileName);
            }

            if (fileName.includes("language-extensions")) {
                const dtsName = fileName.replace(/(\.d)?(\.ts)$/, ".d.ts");
                filePath = path.resolve(dtsName);
            }

            if (filePath !== undefined) {
                if (libCache[fileName]) return libCache[fileName];
                const content = fs.readFileSync(filePath, "utf8");
                libCache[fileName] = ts.createSourceFile(filePath, content, ts.ScriptTarget.Latest, false);
                return libCache[fileName];
            }
        },
    };

    return ts.createProgram(Object.keys(input), options, compilerHost);
}

export interface TranspileVirtualProjectResult {
    diagnostics: ts.Diagnostic[];
    transpiledFiles: TranspiledFile[];
}

export function transpileVirtualProject(
    files: Record<string, string>,
    options: CompilerOptions = {}
): TranspileVirtualProjectResult {
    const program = createVirtualProgram(files, options);
    const collector = createEmitOutputCollector();
    const { diagnostics: transpileDiagnostics } = new Transpiler().emit({ program, writeFile: collector.writeFile });
    const diagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...transpileDiagnostics,
    ]);

    return { diagnostics: [...diagnostics], transpiledFiles: collector.files };
}

export interface TranspileStringResult {
    diagnostics: ts.Diagnostic[];
    file?: TranspiledFile;
}

export function transpileString(main: string, options: CompilerOptions = {}): TranspileStringResult {
    const { diagnostics, transpiledFiles } = transpileVirtualProject({ "main.ts": main }, options);
    return {
        diagnostics,
        file: transpiledFiles.find(({ sourceFiles }) => sourceFiles.some(f => f.fileName === "main.ts")),
    };
}
