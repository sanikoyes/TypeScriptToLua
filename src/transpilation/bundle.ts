import * as path from "path";
import { SourceNode } from "source-map";
import * as ts from "typescript";
import { CompilerOptions } from "../CompilerOptions";
import { escapeString, tstlHeader } from "../LuaPrinter";
import { cast, formatPathToLuaPath, isNonNull, trimExtension } from "../utils";
import { couldNotFindBundleEntryPoint } from "./diagnostics";
import { getEmitOutDir, getEmitPathRelativeToOutDir, getSourceDir } from "./transpiler";
import { EmitFile, ProcessedFile } from "./utils";

const createModulePath = (pathToResolve: string, program: ts.Program) =>
    escapeString(formatPathToLuaPath(trimExtension(getEmitPathRelativeToOutDir(pathToResolve, program))));

// Override `require` to read from ____modules table.
const requireOverride = `
local ____modules = {}
local ____moduleCache = {}
local ____originalRequire = require
local function require(file)
    if ____moduleCache[file] then
        return ____moduleCache[file].value
    end
    if ____modules[file] then
        ____moduleCache[file] = { value = ____modules[file]() }
        return ____moduleCache[file].value
    else
        if ____originalRequire then
            return ____originalRequire(file)
        else
            error("module '" .. file .. "' not found")
        end
    end
end
`;

export function getBundleResult(program: ts.Program, files: ProcessedFile[]): [ts.Diagnostic[], EmitFile] {
    const diagnostics: ts.Diagnostic[] = [];

    const options = program.getCompilerOptions() as CompilerOptions;
    const bundleFile = cast(options.luaBundle, isNonNull);
    const entryModule = cast(options.luaBundleEntry, isNonNull);

    // Resolve project settings relative to project file.
    const resolvedEntryModule = path.resolve(getSourceDir(program), entryModule);
    const outputPath = path.resolve(getEmitOutDir(program), bundleFile);

    if (program.getSourceFile(resolvedEntryModule) === undefined && program.getSourceFile(entryModule) === undefined) {
        diagnostics.push(couldNotFindBundleEntryPoint(entryModule));
    }

    // For each file: ["<module path>"] = function() <lua content> end,
    const moduleTableEntries = files.map(f => moduleSourceNode(f, createModulePath(f.fileName, program)));

    // Create ____modules table containing all entries from moduleTableEntries
    const moduleTable = createModuleTableNode(moduleTableEntries);

    // return require("<entry module path>")
    const entryPoint = `return require(${createModulePath(entryModule, program)})\n`;

    const sourceChunks = [requireOverride, moduleTable, entryPoint];

    if (!options.noHeader) {
        sourceChunks.unshift(tstlHeader);
    }

    const bundleNode = joinSourceChunks(sourceChunks);
    const { code, map } = bundleNode.toStringWithSourceMap();

    return [
        diagnostics,
        {
            outputPath,
            code,
            sourceMap: map.toString(),
            sourceFiles: files.flatMap(x => x.sourceFiles ?? []),
        },
    ];
}

function moduleSourceNode({ code, sourceMapNode }: ProcessedFile, modulePath: string): SourceNode {
    const tableEntryHead = `[${modulePath}] = function() `;
    const tableEntryTail = " end,\n";

    return joinSourceChunks([tableEntryHead, sourceMapNode ?? code, tableEntryTail]);
}

function createModuleTableNode(fileChunks: SourceChunk[]): SourceNode {
    const tableHead = "____modules = {\n";
    const tableEnd = "}\n";

    return joinSourceChunks([tableHead, ...fileChunks, tableEnd]);
}

type SourceChunk = string | SourceNode;
function joinSourceChunks(chunks: SourceChunk[]): SourceNode {
    return new SourceNode(null, null, null, chunks);
}
