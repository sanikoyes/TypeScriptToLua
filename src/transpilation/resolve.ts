import * as path from "path";
import * as resolve from "enhanced-resolve";
import * as ts from "typescript";
import * as fs from "fs";
import { EmitHost, ProcessedFile } from "./utils";
import { SourceNode } from "source-map";
import { getEmitPathRelativeToOutDir, getProjectRoot, getSourceDir } from "./transpiler";
import { formatPathToLuaPath, normalizeSlashes, trimExtension } from "../utils";
import { couldNotReadDependency, couldNotResolveRequire } from "./diagnostics";
import { BuildMode, CompilerOptions } from "../CompilerOptions";

const resolver = resolve.ResolverFactory.createResolver({
    extensions: [".lua"],
    enforceExtension: true, // Resolved file must be a lua file
    fileSystem: { ...new resolve.CachedInputFileSystem(fs) },
    useSyncFileSystemCalls: true,
});

interface ResolutionResult {
    resolvedFiles: ProcessedFile[];
    diagnostics: ts.Diagnostic[];
}

class ResolutionContext {
    private resultsCache = new Map<string, ResolutionResult>();

    constructor(
        public readonly program: ts.Program,
        public readonly options: CompilerOptions,
        private readonly emitHost: EmitHost
    ) {}

    public resolve(file: ProcessedFile, required: string): ResolutionResult {
        const resolvedDependency = resolveDependency(file, required, this.program, this.emitHost);
        if (resolvedDependency) {
            if (this.options.tstlVerbose) {
                console.log(`Resolved ${required} to ${normalizeSlashes(resolvedDependency)}`);
            }

            // Figure out resolved require path and dependency output path
            const resolvedRequire = getEmitPathRelativeToOutDir(resolvedDependency, this.program);

            if (shouldRewriteRequires(resolvedDependency, this.program)) {
                replaceRequireInCode(file, required, resolvedRequire);
                replaceRequireInSourceMap(file, required, resolvedRequire);
            }

            // Check cache to prevent resolving nested dependencies double to break dependency loops
            if (this.resultsCache.has(resolvedDependency)) {
                if (this.options.tstlVerbose) {
                    console.log(`Resolution cache hit for ${normalizeSlashes(resolvedDependency)}`);
                }
                return this.resultsCache.get(resolvedDependency)!;
            }

            // If dependency is not part of project, add dependency to output and resolve its dependencies recursively
            if (shouldIncludeDependency(resolvedDependency, this.program)) {
                // If dependency resolved successfully, read its content
                const dependencyContent = this.emitHost.readFile(resolvedDependency);
                if (dependencyContent === undefined) {
                    return { resolvedFiles: [], diagnostics: [couldNotReadDependency(resolvedDependency)] };
                }

                const dependency = {
                    fileName: resolvedDependency,
                    code: dependencyContent,
                };
                const nestedDependencies = resolveFileDependencies(dependency, this);

                // Cache result and return
                const result = {
                    resolvedFiles: [dependency, ...nestedDependencies.resolvedFiles],
                    diagnostics: [...nestedDependencies.diagnostics],
                };
                this.resultsCache.set(resolvedDependency, result);
                return result;
            } else {
                const result = {
                    resolvedFiles: [],
                    diagnostics: [],
                };
                this.resultsCache.set(resolvedDependency, result);
                return result;
            }
        } else {
            const fallbackRequire = fallbackResolve(required, getSourceDir(this.program), path.dirname(file.fileName));
            replaceRequireInCode(file, required, fallbackRequire);
            replaceRequireInSourceMap(file, required, fallbackRequire);

            return {
                resolvedFiles: [],
                diagnostics: [
                    couldNotResolveRequire(required, path.relative(getProjectRoot(this.program), file.fileName)),
                ],
            };
        }
    }
}

export function resolveDependencies(program: ts.Program, files: ProcessedFile[], emitHost: EmitHost): ResolutionResult {
    const outFiles: ProcessedFile[] = [...files];
    const diagnostics: ts.Diagnostic[] = [];
    const options = program.getCompilerOptions() as CompilerOptions;

    const resolutionContext = new ResolutionContext(program, options, emitHost);

    // Resolve dependencies for all processed files
    for (const file of files) {
        if (options.tstlVerbose) {
            console.log(`Resolving dependencies for ${normalizeSlashes(file.fileName)}`);
        }

        const resolutionResult = resolveFileDependencies(file, resolutionContext);
        outFiles.push(...resolutionResult.resolvedFiles);
        diagnostics.push(...resolutionResult.diagnostics);
    }

    return { resolvedFiles: deduplicateResolvedFiles(outFiles), diagnostics };
}

function deduplicateResolvedFiles(files: ProcessedFile[]): ProcessedFile[] {
    return [...new Map(files.map(f => [f.fileName, f])).values()];
}

function resolveFileDependencies(file: ProcessedFile, context: ResolutionContext): ResolutionResult {
    const dependencies: ProcessedFile[] = [];
    const diagnostics: ts.Diagnostic[] = [];

    for (const required of findRequiredPaths(file.code)) {
        // Do no resolve lualib
        if (required === "lualib_bundle") {
            dependencies.push({ fileName: "lualib_bundle", code: "" });
            continue;
        }

        // Do not resolve noResolution paths
        if (required.startsWith("@NoResolution:")) {
            // Remove @NoResolution prefix if not building in library mode
            if (!isBuildModeLibrary(context.program)) {
                const path = required.replace("@NoResolution:", "");
                replaceRequireInCode(file, required, path);
                replaceRequireInSourceMap(file, required, path);
            }

            // Skip
            continue;
        }

        // Try to resolve the import starting from the directory `file` is in
        const resolvedDependency = context.resolve(file, required);
        dependencies.push(...resolvedDependency.resolvedFiles);
        diagnostics.push(...resolvedDependency.diagnostics);
    }
    return { resolvedFiles: dependencies, diagnostics };
}

function resolveDependency(
    requiringFile: ProcessedFile,
    dependency: string,
    program: ts.Program,
    emitHost: EmitHost
): string | undefined {
    const options = program.getCompilerOptions() as CompilerOptions;

    const fileDirectory = path.dirname(requiringFile.fileName);
    if (options.tstlVerbose) {
        console.log(`Resolving "${dependency}" from ${normalizeSlashes(fileDirectory)}`);
    }

    // Check if file is a file in the project
    const resolvedPath = path.join(options.baseUrl ?? fileDirectory, dependency);

    const possibleProjectFiles = [
        resolvedPath, // JSON files need their extension as part of the import path, caught by this branch,
        resolvedPath + ".ts", // Regular ts file
        path.join(resolvedPath, "index.ts"), // Index ts file,
        resolvedPath + ".tsx", // tsx file
        path.join(resolvedPath, "index.tsx"), // tsx index
    ];

    for (const possibleFile of possibleProjectFiles) {
        if (isProjectFile(possibleFile, program)) {
            return possibleFile;
        }
    }

    // Check if this is a lua file in the project sources
    const possibleLuaProjectFiles = [
        resolvedPath + ".lua", // lua file in sources
        path.join(resolvedPath, "index.lua"), // lua index file in sources
    ];

    for (const possibleFile of possibleLuaProjectFiles) {
        if (emitHost.fileExists(possibleFile)) {
            return possibleFile;
        }
    }

    // Check if this is a sibling of a required lua file
    if (requiringFile.fileName.endsWith(".lua")) {
        const luaFilePath = resolveLuaPath(fileDirectory, dependency, emitHost);
        if (luaFilePath) {
            return luaFilePath;
        }
    }

    // Not a TS file in our project sources, use resolver to check if we can find dependency
    try {
        const resolveResult = resolver.resolveSync({}, fileDirectory, dependency);
        if (resolveResult) {
            return resolveResult;
        }
    } catch (e) {
        // resolveSync errors if it fails to resolve
    }

    return undefined;
}

function resolveLuaPath(fromPath: string, dependency: string, emitHost: EmitHost) {
    const splitDependency = dependency.split(".");
    if (splitDependency.length === 1) {
        // If dependency has just one part (the file), look for a lua file with that name
        const fileDirectory = walkUpFileTreeUntil(fromPath, dir =>
            emitHost.fileExists(path.join(dir, dependency) + ".lua")
        );
        if (fileDirectory) {
            return path.join(fileDirectory, dependency) + ".lua";
        }
    } else {
        // If dependency has multiple parts, look for the first directory of the require path, which must be in the lua root
        const luaRoot = walkUpFileTreeUntil(fromPath, dir =>
            emitHost.directoryExists(path.join(dir, splitDependency[0]))
        );
        if (luaRoot) {
            return path.join(luaRoot, dependency.replace(/\./g, path.sep)) + ".lua";
        }
    }
}

function walkUpFileTreeUntil(fromDirectory: string, predicate: (dir: string) => boolean) {
    const currentDir = path.normalize(fromDirectory).split(path.sep);
    while (currentDir.length > 0) {
        const dir = currentDir.join(path.sep);
        if (predicate(dir)) {
            return dir;
        }
        currentDir.pop();
    }
    return undefined;
}

function shouldRewriteRequires(resolvedDependency: string, program: ts.Program) {
    return !isNodeModulesFile(resolvedDependency) || !isBuildModeLibrary(program);
}

function shouldIncludeDependency(resolvedDependency: string, program: ts.Program) {
    // Never include lua files (again) that are transpiled from project sources
    if (!hasSourceFileInProject(resolvedDependency, program)) {
        // Always include lua files not in node_modules (internal lua sources)
        if (!isNodeModulesFile(resolvedDependency)) {
            return true;
        } else {
            // Only include node_modules files if not in library mode
            return !isBuildModeLibrary(program);
        }
    }
    return false;
}

function isBuildModeLibrary(program: ts.Program) {
    return program.getCompilerOptions().buildMode === BuildMode.Library;
}

function findRequiredPaths(code: string): string[] {
    // Find all require("<path>") paths in a lua code string
    const paths: string[] = [];
    const pattern = /(^|\s|;|=)require\("(.+)"\)/g;
    // eslint-disable-next-line @typescript-eslint/ban-types
    let match: RegExpExecArray | null;
    while ((match = pattern.exec(code))) {
        paths.push(match[2]);
    }

    return paths;
}

function replaceRequireInCode(file: ProcessedFile, originalRequire: string, newRequire: string): void {
    const requirePath = formatPathToLuaPath(newRequire.replace(".lua", ""));

    // Escape special characters to prevent the regex from breaking...
    const escapedRequire = originalRequire.replace(/[-/\\^$*+?.()|[\]{}]/g, "\\$&");

    file.code = file.code.replace(
        new RegExp(`(^|\\s|;|=)require\\("${escapedRequire}"\\)`),
        `$1require("${requirePath}")`
    );
}

function replaceRequireInSourceMap(file: ProcessedFile, originalRequire: string, newRequire: string): void {
    const requirePath = formatPathToLuaPath(newRequire.replace(".lua", ""));
    if (file.sourceMapNode) {
        replaceInSourceMap(file.sourceMapNode, file.sourceMapNode, `"${originalRequire}"`, `"${requirePath}"`);
    }
}

function replaceInSourceMap(node: SourceNode, parent: SourceNode, require: string, resolvedRequire: string): boolean {
    if ((!node.children || node.children.length === 0) && node.toString() === require) {
        parent.children = [new SourceNode(node.line, node.column, node.source, [resolvedRequire])];
        return true; // Stop after finding the first occurrence
    }

    if (node.children) {
        for (const c of node.children) {
            if (replaceInSourceMap(c, node, require, resolvedRequire)) {
                return true; // Occurrence found in one of the children
            }
        }
    }

    return false; // Did not find the require
}

function isNodeModulesFile(filePath: string): boolean {
    return path
        .normalize(filePath)
        .split(path.sep)
        .some(p => p === "node_modules");
}

function isProjectFile(file: string, program: ts.Program): boolean {
    return program.getSourceFile(file) !== undefined;
}

function hasSourceFileInProject(filePath: string, program: ts.Program) {
    const pathWithoutExtension = trimExtension(filePath);
    return (
        isProjectFile(pathWithoutExtension + ".ts", program) ||
        isProjectFile(pathWithoutExtension + ".tsx", program) ||
        isProjectFile(pathWithoutExtension + ".json", program)
    );
}

// Transform an import path to a lua require that is probably not correct, but can be used as fallback when regular resolution fails
function fallbackResolve(required: string, sourceRootDir: string, fileDir: string): string {
    return formatPathToLuaPath(
        path
            .normalize(path.join(path.relative(sourceRootDir, fileDir), required))
            .split(path.sep)
            .filter(s => s !== "." && s !== "..")
            .join(path.sep)
    );
}
