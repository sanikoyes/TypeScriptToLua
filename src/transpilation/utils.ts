import * as path from "path";
import * as resolve from "resolve";
import { SourceNode } from "source-map";
import * as ts from "typescript";
// TODO: Don't depend on CLI?
import * as cliDiagnostics from "../cli/diagnostics";
import * as lua from "../LuaAST";
import * as diagnosticFactories from "./diagnostics";

export interface EmitHost {
    directoryExists(path: string): boolean;
    fileExists(path: string): boolean;
    getCurrentDirectory(): string;
    readFile(path: string): string | undefined;
    writeFile: ts.WriteFileCallback;
}

interface BaseFile {
    code: string;
    sourceMap?: string;
    sourceFiles?: ts.SourceFile[];
}

export interface ProcessedFile extends BaseFile {
    fileName: string;
    luaAst?: lua.File;
    /** @internal */
    sourceMapNode?: SourceNode;
}

export interface EmitFile extends BaseFile {
    outputPath: string;
}

export const getConfigDirectory = (options: ts.CompilerOptions) =>
    options.configFilePath ? path.dirname(options.configFilePath) : process.cwd();

export function resolvePlugin(
    kind: string,
    optionName: string,
    basedir: string,
    query: string,
    importName = "default"
): { error?: ts.Diagnostic; result?: unknown } {
    if (typeof query !== "string") {
        return { error: cliDiagnostics.compilerOptionRequiresAValueOfType(optionName, "string") };
    }

    let resolved: string;
    try {
        resolved = resolve.sync(query, { basedir, extensions: [".js", ".ts", ".tsx"] });
    } catch (err) {
        if (err.code !== "MODULE_NOT_FOUND") throw err;
        return { error: diagnosticFactories.couldNotResolveFrom(kind, query, basedir) };
    }

    const hasNoRequireHook = require.extensions[".ts"] === undefined;
    if (hasNoRequireHook && (resolved.endsWith(".ts") || resolved.endsWith(".tsx"))) {
        try {
            const tsNodePath = resolve.sync("ts-node", { basedir });
            const tsNode: typeof import("ts-node") = require(tsNodePath);
            tsNode.register({ transpileOnly: true });
        } catch (err) {
            if (err.code !== "MODULE_NOT_FOUND") throw err;
            return { error: diagnosticFactories.toLoadItShouldBeTranspiled(kind, query) };
        }
    }

    const commonjsModule = require(resolved);
    const factoryModule = commonjsModule.__esModule ? commonjsModule : { default: commonjsModule };
    const result = factoryModule[importName];
    if (result === undefined) {
        return { error: diagnosticFactories.shouldHaveAExport(kind, query, importName) };
    }

    return { result };
}
