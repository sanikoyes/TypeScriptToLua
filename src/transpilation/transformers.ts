import * as ts from "typescript";
// TODO: Don't depend on CLI?
import * as cliDiagnostics from "../cli/diagnostics";
import { CompilerOptions, TransformerImport } from "../CompilerOptions";
import * as diagnosticFactories from "./diagnostics";
import { getConfigDirectory, resolvePlugin } from "./utils";

export function getTransformers(
    program: ts.Program,
    diagnostics: ts.Diagnostic[],
    customTransformers: ts.CustomTransformers,
    onSourceFile: (sourceFile: ts.SourceFile) => void
): ts.CustomTransformers {
    const luaTransformer: ts.TransformerFactory<ts.SourceFile> = () => sourceFile => {
        onSourceFile(sourceFile);
        return ts.createSourceFile(sourceFile.fileName, "", ts.ScriptTarget.ESNext);
    };

    const transformersFromOptions = loadTransformersFromOptions(program, diagnostics);

    const afterDeclarations = [
        ...(transformersFromOptions.afterDeclarations ?? []),
        ...(customTransformers.afterDeclarations ?? []),
    ];

    const options = program.getCompilerOptions() as CompilerOptions;
    if (options.noImplicitSelf) {
        afterDeclarations.unshift(noImplicitSelfTransformer);
    }

    return {
        afterDeclarations,
        before: [
            ...(customTransformers.before ?? []),
            ...(transformersFromOptions.before ?? []),

            ...(transformersFromOptions.after ?? []),
            ...(customTransformers.after ?? []),
            stripParenthesisExpressionsTransformer,
            luaTransformer,
        ],
    };
}

export const noImplicitSelfTransformer: ts.TransformerFactory<ts.SourceFile | ts.Bundle> = () => node => {
    const transformSourceFile: ts.Transformer<ts.SourceFile> = node => {
        const empty = ts.factory.createNotEmittedStatement(undefined!);
        ts.addSyntheticLeadingComment(empty, ts.SyntaxKind.MultiLineCommentTrivia, "* @noSelfInFile ", true);
        return ts.factory.updateSourceFile(node, [empty, ...node.statements], node.isDeclarationFile);
    };

    return ts.isBundle(node)
        ? ts.factory.updateBundle(node, node.sourceFiles.map(transformSourceFile))
        : transformSourceFile(node);
};

export const stripParenthesisExpressionsTransformer: ts.TransformerFactory<ts.SourceFile> = context => sourceFile => {
    // Remove parenthesis expressions before transforming to Lua, so transpiler is not hindered by extra ParenthesizedExpression nodes
    function unwrapParentheses(node: ts.Expression) {
        while (ts.isParenthesizedExpression(node)) {
            node = node.expression;
        }
        return node;
    }
    function visit(node: ts.Node): ts.Node {
        // For now only call expressions strip their expressions of parentheses, there could be more cases where this is required
        if (ts.isCallExpression(node)) {
            return ts.factory.updateCallExpression(
                node,
                unwrapParentheses(node.expression),
                node.typeArguments,
                node.arguments
            );
        }

        return ts.visitEachChild(node, visit, context);
    }
    return ts.visitNode(sourceFile, visit);
};

function loadTransformersFromOptions(program: ts.Program, diagnostics: ts.Diagnostic[]): ts.CustomTransformers {
    const customTransformers: Required<ts.CustomTransformers> = {
        before: [],
        after: [],
        afterDeclarations: [],
    };

    const options = program.getCompilerOptions() as CompilerOptions;
    if (!options.plugins) return customTransformers;

    for (const [index, transformerImport] of options.plugins.entries()) {
        if (!("transform" in transformerImport)) continue;
        const optionName = `compilerOptions.plugins[${index}]`;

        const { error: resolveError, result: factory } = resolvePlugin(
            "transformer",
            `${optionName}.transform`,
            getConfigDirectory(options),
            transformerImport.transform,
            transformerImport.import
        );

        if (resolveError) diagnostics.push(resolveError);
        if (factory === undefined) continue;

        const { error: loadError, transformer } = loadTransformer(optionName, program, factory, transformerImport);
        if (loadError) diagnostics.push(loadError);
        if (transformer === undefined) continue;

        if (transformer.before) {
            customTransformers.before.push(transformer.before);
        }

        if (transformer.after) {
            customTransformers.after.push(transformer.after);
        }

        if (transformer.afterDeclarations) {
            customTransformers.afterDeclarations.push(transformer.afterDeclarations);
        }
    }

    return customTransformers;
}

type ProgramTransformerFactory = (program: ts.Program, options: Record<string, any>) => Transformer;
type ConfigTransformerFactory = (options: Record<string, any>) => Transformer;
type CompilerOptionsTransformerFactory = (
    compilerOptions: CompilerOptions,
    options: Record<string, any>
) => Transformer;
type TypeCheckerTransformerFactory = (typeChecker: ts.TypeChecker, options: Record<string, any>) => Transformer;
type RawTransformerFactory = Transformer;

type Transformer = GroupTransformer | ts.TransformerFactory<ts.SourceFile>;
interface GroupTransformer {
    before?: ts.TransformerFactory<ts.SourceFile>;
    after?: ts.TransformerFactory<ts.SourceFile>;
    afterDeclarations?: ts.TransformerFactory<ts.SourceFile | ts.Bundle>;
}

function loadTransformer(
    optionPath: string,
    program: ts.Program,
    factory: unknown,
    { transform, after = false, afterDeclarations = false, type = "program", ...extraOptions }: TransformerImport
): { error?: ts.Diagnostic; transformer?: GroupTransformer } {
    let transformer: Transformer;
    switch (type) {
        case "program":
            transformer = (factory as ProgramTransformerFactory)(program, extraOptions);
            break;
        case "config":
            transformer = (factory as ConfigTransformerFactory)(extraOptions);
            break;
        case "checker":
            transformer = (factory as TypeCheckerTransformerFactory)(program.getTypeChecker(), extraOptions);
            break;
        case "raw":
            transformer = factory as RawTransformerFactory;
            break;
        case "compilerOptions":
            transformer = (factory as CompilerOptionsTransformerFactory)(program.getCompilerOptions(), extraOptions);
            break;
        default: {
            const optionName = `--${optionPath}.type`;
            return { error: cliDiagnostics.argumentForOptionMustBe(optionName, "program") };
        }
    }

    if (typeof after !== "boolean") {
        const optionName = `${optionPath}.after`;
        return { error: cliDiagnostics.compilerOptionRequiresAValueOfType(optionName, "boolean") };
    }

    if (typeof afterDeclarations !== "boolean") {
        const optionName = `${optionPath}.afterDeclarations`;
        return { error: cliDiagnostics.compilerOptionRequiresAValueOfType(optionName, "boolean") };
    }

    if (typeof transformer === "function") {
        let wrappedTransformer: GroupTransformer;

        if (after) {
            wrappedTransformer = { after: transformer };
        } else if (afterDeclarations) {
            wrappedTransformer = { afterDeclarations: transformer as ts.TransformerFactory<ts.SourceFile | ts.Bundle> };
        } else {
            wrappedTransformer = { before: transformer };
        }

        return { transformer: wrappedTransformer };
    } else {
        const isValidGroupTransformer =
            typeof transformer === "object" &&
            (transformer.before ?? transformer.after ?? transformer.afterDeclarations) !== undefined;

        if (!isValidGroupTransformer) {
            return { error: diagnosticFactories.transformerShouldBeATsTransformerFactory(transform) };
        }
    }

    return { transformer };
}
