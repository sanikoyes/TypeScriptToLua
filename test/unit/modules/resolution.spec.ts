import * as ts from "typescript";
import { couldNotResolveRequire } from "../../../src/transpilation/diagnostics";
import * as util from "../../util";

const requireRegex = /require\("(.*?)"\)/;
const expectToRequire =
    (expected: string): util.TapCallback =>
    builder => {
        const [, requiredPath] = builder.getMainLuaCodeChunk().match(requireRegex) ?? [];
        expect(requiredPath).toBe(expected);
    };

test.each([
    {
        filePath: "main.ts",
        usedPath: "./folder/Module",
        expected: "folder.Module",
        options: { rootDir: "." },
    },
    {
        filePath: "main.ts",
        usedPath: "./folder/Module",
        expected: "folder.Module",
        options: { rootDir: "./" },
    },
    {
        filePath: "src/main.ts",
        usedPath: "./folder/Module",
        expected: "src.folder.Module",
        options: { rootDir: "." },
    },
    {
        filePath: "main.ts",
        usedPath: "folder/Module",
        expected: "folder.Module",
        options: { rootDir: ".", baseUrl: "." },
    },
    {
        filePath: "main.ts",
        usedPath: "folder/Module",
        expected: "folder.Module",
        options: { rootDir: "./", baseUrl: "." },
    },
    {
        filePath: "src/main.ts",
        usedPath: "./folder/Module",
        expected: "folder.Module",
        options: { rootDir: "src" },
    },
    {
        filePath: "src/main.ts",
        usedPath: "./folder/Module",
        expected: "folder.Module",
        options: { rootDir: "./src" },
    },
    {
        filePath: "src/dir/main.ts",
        usedPath: "../Module",
        expected: "Module",
        options: { rootDir: "./src" },
    },
    {
        filePath: "src/dir/dir/main.ts",
        usedPath: "../../dir/Module",
        expected: "dir.Module",
        options: { rootDir: "./src" },
    },
])("resolve paths with baseUrl or rootDir (%p)", ({ filePath, usedPath, expected, options }) => {
    util.testModule`
        import * as module from "${usedPath}";
        module;
    `
        .setMainFileName(filePath)
        .addExtraFile(`${usedPath}.ts`, "")
        .setOptions(options)
        .tap(expectToRequire(expected));
});

test("doesn't resolve paths out of root dir", () => {
    util.testModule`
        import * as module from "../module";
        module;
    `
        .setMainFileName("src/main.ts")
        .setOptions({ rootDir: "./src" })
        .disableSemanticCheck()
        .expectDiagnosticsToMatchSnapshot([couldNotResolveRequire.code]);
});

test.each([
    {
        declarationStatement: `
            /** @noResolution */
            declare module "fake" {}
        `,
        mainCode: 'import "fake";',
        expectedPath: "fake",
    },
    {
        declarationStatement: `
            /** @noResolution */
            declare module "fake" {}
        `,
        mainCode: 'import * as fake from "fake"; fake;',
        expectedPath: "fake",
    },
    {
        declarationStatement: `
            /** @noResolution */
            declare module "fake" {
                export const x: number;
            }
        `,
        mainCode: 'import { x } from "fake"; x;',
        expectedPath: "fake",
    },
    {
        declarationStatement: `
            /** @noResolution */
            declare module "fake" {
                export const x: number;
            }

            declare module "fake" {
                export const y: number;
            }
        `,
        mainCode: 'import { y } from "fake"; y;',
        expectedPath: "fake",
    },
])("noResolution prevents any module path resolution behavior", ({ declarationStatement, mainCode, expectedPath }) => {
    util.testModule(mainCode)
        .setMainFileName("src/main.ts")
        .addExtraFile("module.d.ts", declarationStatement)
        .tap(expectToRequire(expectedPath));
});

test("import = require", () => {
    util.testModule`
        import foo = require("./foo/bar");
        foo;
    `
        .setOptions({ module: ts.ModuleKind.CommonJS })
        .tap(expectToRequire("foo.bar"));
});
