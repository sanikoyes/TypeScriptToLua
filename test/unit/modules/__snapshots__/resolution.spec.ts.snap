// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't resolve paths out of root dir: code 1`] = `
"local ____exports = {}
local module = require(\\"module\\")
local ____ = module
return ____exports"
`;

exports[`doesn't resolve paths out of root dir: diagnostics 1`] = `"error TSTL: Could not resolve require path '../module' in file main.ts."`;
